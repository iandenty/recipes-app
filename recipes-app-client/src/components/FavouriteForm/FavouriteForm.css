.favouite__button {
  color:#565656;
  font-size: 3rem;
  line-height: 3rem;
  border: none;
  padding: 2px;
  background: none;
  cursor: pointer;
}

.favouite__button--active {
  color: gold;
}