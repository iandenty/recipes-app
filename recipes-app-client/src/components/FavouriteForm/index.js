import React, { Component } from 'react';
import PropTypes from 'prop-types';

import './FavouriteForm.css';

class FavouriteForm extends Component {
  state = {
    isFavourite: false
  }

  componentDidMount() {
    const isSet = localStorage.getItem('favourites');
    if(isSet) {
      const hasFavourited = localStorage.getItem('favourites')
        .split(',')
        .map(favourite => parseInt(favourite));

      if(hasFavourited.includes(this.props.id)) {
        this.setState(prevState => ({
          isFavourite: true
        })); 
      }
    }
  }

  // Using local storage to simulate user experience. Ideally would save to database using
  // action specified below.
  // TODO: Refactor into resusable functions which can be shared with componentDidMount logic.
  storeFavourited(favouritedId) {
    const isSet = localStorage.getItem('favourites');

    if(isSet) {
      const favourites = localStorage.getItem('favourites')
        .split(',')
        .map(favourite => parseInt(favourite));

      const favouriteIndex = favourites.findIndex(favourite => favourite === favouritedId);

      if(favouriteIndex === -1) {
        favourites.push(favouritedId);
        localStorage.setItem('favourites', favourites.join(','))
      } else {
        localStorage.setItem('favourites', favourites.filter(favourite => favourite !== favouritedId));
      }
    } else {
      localStorage.setItem('favourites', favouritedId);
    }
    this.setState(prevState => ({
      isFavourite: !prevState.isFavourite
    }));
  }

  handleSubmit(event) {
    event.preventDefault();
    const favouritedId = parseInt(event.target.favourite.value);
    this.storeFavourited(favouritedId)
  }

  render() {
    const { 
      id
    } = this.props;

    // Represents endpoint to store favourited recipes. This app assumes user exists and is logged in.
    const action = `/api/recipes/${id}/favourite`;

    return (
      <div>
      { this.props.isForm ?
        <form 
          className="favourite"
          method="POST" 
          action={action}
          onSubmit={(e) => this.handleSubmit(e)} 
        >
          <button
            type="submit" 
            name="favourite"
            value={id}
            className={`favouite__button ${this.state.isFavourite ? 'favouite__button--active' : ''}`}
          >☆</button>
        </form>
      :
        <div className="favourite">
          <span
            className={`favouite__button ${this.state.isFavourite ? 'favouite__button--active' : ''}`}
          >☆</span>
        </div>
      }
      </div>
    );
  }
}

FavouriteForm.defaultProps = {
  isForm: true
}

FavouriteForm.propTypes = {
  id: PropTypes.number
};

export default FavouriteForm;
