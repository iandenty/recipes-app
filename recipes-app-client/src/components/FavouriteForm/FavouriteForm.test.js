import React from 'react';
import { shallow } from 'enzyme';
import FavouriteForm from '../FavouriteForm';
import renderer from 'react-test-renderer';
import 'mock-local-storage';

describe('Favourite Form', () => { 
  const component = shallow(
    <FavouriteForm id={1} />
  );
  const handleSubmit = jest.fn({event: {target: {favourite: {value: true}}}});

  // Snapshot
  it('renders and matches the snapshot', () => {
    const component = renderer.create(
      <FavouriteForm id={1} onSubmit={handleSubmit} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  // Contains provided content
  it('contains the submit button', () => {
    expect(component.find('.favouite__button')).toHaveLength(1);
  });

  // Interactions
  it('calls', () => {
    component.instance().storeFavourited(1) ;
    expect(component.state("isFavourite")).toEqual(true);
  });
});
