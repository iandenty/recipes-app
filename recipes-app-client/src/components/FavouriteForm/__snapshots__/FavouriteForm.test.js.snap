// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favourite Form renders and matches the snapshot 1`] = `
<form
  action="/api/recipes/1/favourite"
  className="favourite"
  method="POST"
  onSubmit={[Function]}
>
  <button
    className="favouite__button "
    name="favourite"
    type="submit"
    value={1}
  >
    <span
      className="visually-hidden"
    >
      Favourite this recipe!
    </span>
    â˜†
  </button>
</form>
`;
